### PR TITLE
Deprecated several functions at torch.nn.functional

### LINKED ISSUE #6245: Deprecate torch.nn.functional.tanh?
@soumith: "we should now deprecate torch.nn.functional.tanh on master, as tensors and variables are now merged."

https://discuss.pytorch.org/t/torch-tanh-vs-torch-nn-functional-tanh/15897

### PR DESCRIPTION
**Pull request by weiyangfb**
## summary
1. fixes #6245
2. deprecated tanh, sigmoid

### CODE REVIEW

File: torch/distributions/utils.py [hunk 1]
```diff
@@ -116,7 +116,7 @@ def logits_to_probs(logits, is_binary=False):
     the log probabilities (possibly unnormalized) of the events.
     """
     if is_binary:
-        return F.sigmoid(logits)
+        return torch.nn.Sigmoid(logits)
```
Comment (soumith) [inline torch/distributions/utils.py]: isn't this `torch.sigmoid(logits)`

File: torch/distributions/transforms.py [hunk 2]
```diff
@@ -497,7 +497,7 @@ def _inverse(self, y):
 
     def log_abs_det_jacobian(self, x, y):
         offset = (x.shape[-1] + 1) - x.new([1]).expand(x.shape).cumsum(-1)
-        z = sigmoid(x - offset.log())
+        z = torch.nn.Sigmoid(x - offset.log())
```
Comment (soumith) [inline torch/distributions/transforms.py]: isn't this `torch.sigmoid(logits)`

File: torch/distributions/transforms.py [hunk 3]
```diff
@@ -483,7 +483,7 @@ def __eq__(self, other):
 
     def _call(self, x):
         offset = (x.shape[-1] + 1) - x.new([1]).expand(x.shape).cumsum(-1)
-        z = sigmoid(x - offset.log())
+        z = torch.nn.Sigmoid(x - offset.log())
```
Comment (soumith) [inline torch/distributions/transforms.py]: isn't this `torch.sigmoid(logits)`

File: torch/distributions/transforms.py [hunk 4]
```diff
@@ -341,7 +341,7 @@ def __eq__(self, other):
         return isinstance(other, SigmoidTransform)
 
     def _call(self, x):
-        return sigmoid(x)
+        return torch.nn.Sigmoid(x)
```
Comment (soumith) [inline torch/distributions/transforms.py]: isn't this `torch.sigmoid(logits)`

File: test/test_jit.py [hunk 5]
```diff
@@ -3777,7 +3777,7 @@ def reparameterize(self, mu, logvar):
 
             def decode(self, z):
                 h3 = F.relu(self.fc3(z))
-                return F.sigmoid(self.fc4(h3))
+                return torch.nn.Sigmoid(self.fc4(h3))
```
Comment (soumith) [inline test/test_jit.py]: isn't this `torch.sigmoid(logits)`

File: test/test_jit.py [hunk 6]
```diff
@@ -61,10 +61,10 @@ def LSTMCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
     gates = F.linear(input, w_ih, b_ih) + F.linear(hx, w_hh, b_hh)
 
     ingate, forgetgate, cellgate, outgate = gates.chunk(4, 1)
-    ingate = F.sigmoid(ingate)
-    forgetgate = F.sigmoid(forgetgate)
+    ingate = torch.nn.Sigmoid(ingate)
+    forgetgate = torch.nn.Sigmoid(forgetgate)
     cellgate = F.tanh(cellgate)
-    outgate = F.sigmoid(outgate)
+    outgate = torch.nn.Sigmoid(outgate)
```
Comment (soumith) [inline test/test_jit.py]: isn't this `torch.sigmoid(logits)`

File: test/test_jit.py [hunk 7]
```diff
@@ -61,10 +61,10 @@ def LSTMCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
     gates = F.linear(input, w_ih, b_ih) + F.linear(hx, w_hh, b_hh)
 
     ingate, forgetgate, cellgate, outgate = gates.chunk(4, 1)
-    ingate = F.sigmoid(ingate)
-    forgetgate = F.sigmoid(forgetgate)
+    ingate = torch.nn.Sigmoid(ingate)
```
Comment (soumith) [inline test/test_jit.py]: isn't this `torch.sigmoid(logits)`

File: test/test_jit.py (hunk 6 above)
Comment (soumith) [inline test/test_jit.py]: isn't this `torch.sigmoid(logits)`

**Review by soumith — CHANGES_REQUESTED** [review]

**Comment by soumith** [comment]
what happens if user tries to do `torch.nn.functional.sigmoid(x)` now, is it an error or a deprecated warning? (it should be a deprecated warning)

**Comment by weiyangfb** [comment]
@soumith ah, I see. Now it is an error, I will make a deprecated warning

**Comment by vishwakftw** [comment]
Other functions in `torch.nn.functional` that could be deprecated:
1. `selu`
2. `rrelu`

Let me know what you think @soumith , I could open a PR for those.

**Comment by weiyangfb** [comment]
@vishwakftw Yes I can deprecate `relu` and `rrelu`. But after that people can only use modules `nn.ReLU` and `nn.RReLU`. Should we do this? 

cc @soumith

**Comment by soumith** [comment]
wait, why would we deprecate `relu` and `rrelu`? we dont have `torch.relu` and `torch.rrelu`

**Comment by weiyangfb** [comment]
@soumith That's right. We will keep them for now then.

**Comment by vishwakftw** [comment]
I thought we could deprecate `torch.nn.functional.rrelu` and `torch.nn.functional.selu`. Both are available as `torch.rrelu` and `torch.selu` (and their inplace versions).

**Comment by weiyangfb** [comment]
@vishwakftw Ah, I see! So they are just not documented yet, I will add them to the docs. This way I think `functional.selu` and `functional.rrelu` can be safely deprecated then.

```
input = torch.randn(2)
torch.selu(input)
```
tensor([1.4261, 0.2250])

```
input = torch.randn(2)
torch.rrelu(input)
```
tensor([-0.0398, -0.1174])

**Comment by fmassa** [comment]
So the goal now is to expose everything in the `torch` namespace?

**Comment by weiyangfb** [comment]
After offline discussion, we decide to move tanh and sigmoid from nn.functional to torch since they are more of common scientific functions, and keep the others at nn.functional for now.

**Comment by weiyangfb** [comment]
Failed Caffe2 CI tests look not related. Can I get a stamp on this?

**Review by soumith — APPROVED** [review]

**Comment by vadimkantorov** [comment]
`relu` also already is available on `torch` namespace: ` torch.relu(torch.Tensor([1]))` works

**Comment by soumith** [comment]
@vadimkantorov we probably will remove that, please dont rely on undocumented features. the `nn` namespace is really the place for neural-net specifi functions

**Comment by vadimkantorov** [comment]
@soumith ok :) I was hesitating if it was a deficit of documentation or a blunder of auto bindings :)

**Comment by soumith** [comment]
the latter :)
